rust_src/src/casetab.rs

```diff
@@ -3,31 +3,185 @@
+            set_char_table_extras(table, 1, canon);
+            map_char_table(Some(set_canon), Qnil, table, table);
+        }
+
+        if eqv.is_nil() {
+            eqv = Fmake_char_table(Qcase_table, Qnil);
+            map_char_table(Some(set_identity), Qnil, canon, eqv);
+            map_char_table(Some(shuffle), Qnil, canon, eqv);
+            set_char_table_extras(table, 2, eqv);
+        }
+
+        // This is so set_image_of_range_1 in regex.c can find the EQV table.
+        set_char_table_extras(canon, 2, eqv);
+    }
+
+    if standard {
+        unsafe {
+            Vascii_downcase_table = table;
+            Vascii_upcase_table = up;
+            Vascii_canon_table = canon;
+            Vascii_eqv_table = eqv;
+        }
+    } else {
+        let mut buffer: LispBufferRef = current_buffer().into();
+        buffer.downcase_table_ = table;
+        buffer.upcase_table_ = up;
+        buffer.case_canon_table_ = canon;
+        buffer.case_eqv_table_ = eqv;
+    }
+
+    table
+}
+
+// The following functions are called in map_char_table.
+
+// Set CANON char-table element for characters in RANGE to a
+// translated ELT by UP and DOWN char-tables.  This is done only when
+// ELT is a character.  The char-tables CANON, UP, and DOWN are in
+// TABLE.
+extern "C" fn set_canon(table: LispObject, range: LispObject, elt: LispObject) {
+    if let Some(idx) = elt.as_natnum() {
+        let case_table: LispCaseTable = table.into();
+
+        let (up, canon, _) = case_table.extras();
+
+        let up_table: LispCharTableRef = up.into();
+        let value: EmacsInt = up_table.get(idx as isize).into();
+
+        unsafe {
+            Fset_char_table_range(canon, range, case_table.get(value as isize));
+        }
+    }
+}
+
+// Set elements of char-table TABLE for C to C itself.  C may be a
+// cons specifying a character range.  In that case, set characters in
+// that range to themselves.  This is done only when ELT is a
+// character.  This is called in map_char_table.
+extern "C" fn set_identity(table: LispObject, c: LispObject, elt: LispObject) {
+    if elt.is_natnum() {
+        let char_table: LispCharTableRef = table.into();
+
+        let (from, to): (EmacsInt, EmacsInt) = match c.into() {
+            Some((car, cdr)) => (car.into(), cdr.into()),
+            None => {
+                let x = c.into();
+                (x, x)
+            }
+        };
+
+        for i in from..=to {
+            char_table.set_unchecked(i as isize, i.into());
+        }
+    }
+}
+
+// Permute the elements of TABLE (which is initially an identity
+// mapping) so that it has one cycle for each equivalence class
+// induced by the translation table on which map_char_table is
+// operated.
+extern "C" fn shuffle(table: LispObject, c: LispObject, elt: LispObject) {
+    if let Some(idx) = elt.as_natnum() {
+        let char_table: LispCharTableRef = table.into();
+
+        let (from, to): (EmacsInt, EmacsInt) = match c.into() {
```

Comment:
So what is `c` here? A `LispCons` or an `EmacsInt`?

Reply:
Inclusive or

Reply:
It is either sadly.

Reply:
I went to rename it and could not give it a better name than "value". Sigh.